File: filesystem.ts
//% weight=5 color=#002050 icon="\uf0a0"
namespace filesystem {

    /*

    //-% blockId="open_path" block="file %filename"
    //-% weight=80
    export function open(path: string): FileHandle {
        const file = new FileHandle(path);
        file.open();

        return file;
    }

    //-% blockId="flush_file" block="flush %handle"
    export function flush(handle: FileHandle) {
        filesystem.fsFlush(handle.handle);
    }

    //-% blockId="close_file" block="close %handle"
    export function close(handle: FileHandle) {
        filesystem.fsClose(handle.handle);
    }

    //-% blockId="erase_file" block="delete %handle"
    export function erase(handle: FileHandle) {
        filesystem.remove(handle.path);
    }

    //-% blockId="seek_file" block="seek %handle to %position %seekType"
    export function seek(handle: FileHandle, position: number, seekType: SeekType) {
        filesystem.fsSeek(handle.handle, position, seekType);
    }

     */
/*
    export enum SeekType {
        //% block="from the current position"
        CurrentPosition = 0,
        //% block="from the start of the file"
        SetPosition = 1,
        //% block="from the end of the file"
        FileEndOffset = 2
    }

    export enum FileWriteOption {
        //% block="without a new line"
        None = 0,
        //% block="with a new line"
        NewLine = 1
    }

    //% blockId="files_create_handle" block="file %path"
    export function createHandle(path: string) {
        return new FileHandle(path);
    }

    //%
    export class FileHandle {

        path: string;
        handle: number;

        constructor(path: string) {
            this.path = path;
        }

        //% blockId="files_open" block="open %this"
        public open() {
            if (this.handle == -1) {
                this.handle = filesystem.fsOpen(this.path);
            }
        }

        //% blockId="files_flush" block="flush %this"
        public flush() {
            if (this.handle != -1) {
                filesystem.fsClose(this.handle);
            }
        }

        //% blockId="files_close" block="close %this"
        public close() {
            if (this.handle != -1) {
                filesystem.fsClose(this.handle);
            }
        }

        //% blockId="files_erase" block="delete %this"
        public erase() {
            filesystem.remove(this.path);
        }

        //% blockId="files_seek" block="seek %this to %position %seekType"
        public seek(position: number, seekType: SeekType) {
            filesystem.fsSeek(this.handle, position, seekType);
        }

        //% blockId="files_write_string" block="to %this write string %text %newline"
        //% text.shadowOptions.toString=true
        public writeText(text: string, newline: FileWriteOption) {
            filesystem.fsWriteString(this.handle, text + (newline == FileWriteOption.NewLine ? "\r\n" : ""));
        }

        //% blockId="files_read_all" block="read contents of %this"
        public readAll(): string {
            return filesystem.readFile(this.handle);
        }

        //% blockId="files_read_all_lines" block="read lines of %this"
        public readAllLines(): string[] {
            return filesystem.readFile(this.handle).split("\r\n");
        }
    }*/

    //%
    export enum FileOpenFlag {
        //% block="read"
        Read = 0x01,
        //% block="write"
        Write = 0x02,
        //% block="create"
        Create = 0x04
    }

    //% blockId="files_read_string" block="read string from %file"
    export function readString(path: string): string {
        const handle = filesystem.fsOpen(path, FileOpenFlag.Read);
        const buffer = filesystem.fsReadBuffer(handle, 4096);

        filesystem.fsClose(handle);

        return buffer.toString();
    }

    //% blockId="files_read_number" block="read number from %file"
    export function readNumber(path: string): number {
        return parseFloat(filesystem.readString(path));
    }

    //% blockId="files_read_lines" block="read lines of %file"
    export function readLines(path: string): string[] {
        return filesystem.readString(path).split("\r\n");
    }

    //% blockId="files_read_numbers" block="read lines of %file as numbers"
    export function readNumbers(path: string): number[] {
        return filesystem.readLines(path).map(line => parseFloat(line));
    }
}

File: filesystem.cpp
#include "pxt.h"

#if MICROBIT_CODAL
#include "CodalFS.h"
#endif

/**
 * Reading and writing to the filesystem.
 */
//%
namespace filesystem {

// Initializes file system. Must be called before any FS operation.
// built-in size computation for file system
// does not take into account size changes
// for compiled code
    int initFileSystem()
    {
    #if MICROBIT_CODAL
        uBit.filesystem.init();
    #endif
    return 1;
    }

/**
    * Appends text and a new line to a file
    * @param filename file name, eg: "output.txt"
    * @param text the string to append to the end of the file
    */
//% blockId="files_append_line" block="append %text to file %filename"
//% blockExternalInputs=1 weight=90 blockGap=8
//% text.shadowOptions.toString=true
    void appendLine(String text, String filename)
    {
    #if MICROBIT_CODAL
        initFileSystem();

        auto fd = uBit.filesystem.open(MSTR(filename).toCharArray(), FS_CREAT | FS_APPEND | FS_WRITE);
        uBit.filesystem.write(fd, (uint8_t *) MSTR(text).toCharArray(), MSTR(text).length());
        uBit.filesystem.write(fd, (uint8_t *) "\r\n", 2);
        uBit.filesystem.close(fd);
        #endif
    }

/**
    * Appends text to a file
    * @param filename file name, eg: "output.txt"
    * @param text the string to append to the end of the file
    */
//% blockId="fs_append_string" block="append %text to file %filename with newline"
//% blockExternalInputs=1 weight=86 blockGap=8
//% text.shadowOptions.toString=true
    void appendString(String text, String filename)
    {
    #if MICROBIT_CODAL
        initFileSystem();

        auto fd = uBit.filesystem.open(MSTR(filename).toCharArray(), FS_CREAT | FS_APPEND | FS_WRITE);
        uBit.filesystem.write(fd, (uint8_t *) MSTR(text).toCharArray(), MSTR(text).length());
        uBit.filesystem.close(fd);
        #endif
    }

/**
* Reads the content of the file to send it to serial
* @param filename file name, eg: "output.txt"
*/
//% blockId="fs_write_to_serial" block="read file %filename to serial"
//% weight=80
    void readToSerial(String filename) {
    #if MICROBIT_CODAL
        initFileSystem();

        auto fd = uBit.filesystem.open(MSTR(filename).toCharArray(), FS_READ);

        char buf[32];

        int read = 0;
        while((read = uBit.filesystem.read(fd, (uint8_t *) buf, sizeof(buf))) > 0) {
            uBit.serial.send((uint8_t*)buf, read * sizeof(char), SYNC_SPINWAIT);
        }

        uBit.filesystem.close(fd);
        #endif
    }

    /**
    * Reads the contents of the file
    * @param filename file name, eg: "output.txt"
    */
    //% blockId="fs_read_file" block="read file %fd"
    //% weight=80 advanced=true
        String readFile(int fd) {
        #if MICROBIT_CODAL
            initFileSystem();

            char buf[128];

            int read = 0;
            while((read = uBit.filesystem.read(fd, (uint8_t *) buf, sizeof(buf))) > 0) {

            }

            uBit.filesystem.close(fd);

            return mkString(buf, sizeof(buf));
            #endif
        }

/**
    * Removes the file. There is no undo for this operation.
    * @param filename name of the file to remove, eg: "output.txt"
    */
//% blockId="fs_remove" block="delete file %filename"
//% weight=80 blockGap=8
    void remove(String filename)
    {
    #if MICROBIT_CODAL
        initFileSystem();
        uBit.filesystem.remove(MSTR(filename).toCharArray());
        #endif
    }

/**
* Creates a directory
* @param name full qualified path to the new directory
*/
//% weight=10
//% blockId=files_create_directory block="create directory %name"
    void createDirectory(String name) {
    #if MICROBIT_CODAL
        initFileSystem();
        uBit.filesystem.createDirectory(MSTR(name).toCharArray());
        #endif
    }

/**
*
*/
//% blockId="fs_open" block="open file %path with flags %flags"
    int fsOpen(String path, int flags) {
    #if MICROBIT_CODAL
        initFileSystem();
        return uBit.filesystem.open(MSTR(path).toCharArray(), flags);
        #endif
    }

/**
*
*/
//% blockId="fs_flush" block="flush file handle %fd" advanced=true
    int fsFlush(int fd) {
    #if MICROBIT_CODAL
        if (fd < 0) return MICROBIT_NOT_SUPPORTED;

        initFileSystem();
        return uBit.filesystem.flush(fd);
        #endif
    }

/**
*
*/
//% blockId="fs_close" block="close file handle %fd" advanced=true
    int fsClose(int fd) {
    #if MICROBIT_CODAL
        if (fd < 0) return MICROBIT_NOT_SUPPORTED;

        initFileSystem();
        return uBit.filesystem.close(fd);
        #endif
    }

/**
*
*/
//% blockId="fs_seek" block="seek file %fd to offset %offset with flags %flags" advanced=true
    int fsSeek(int fd, int offset, int flags) {
    #if MICROBIT_CODAL
        if (fd < 0) return MICROBIT_NOT_SUPPORTED;
        if (offset < 0) return MICROBIT_INVALID_PARAMETER;

        initFileSystem();
        return uBit.filesystem.seek(fd, offset, flags);
        #endif
    }

/**
*
*/
//% weight=0 advanced=true
    int fsWriteString(int fd, String text) {
    #if MICROBIT_CODAL
        if (fd < 0) return MICROBIT_NOT_SUPPORTED;

        initFileSystem();

        return uBit.filesystem.write(fd, (uint8_t *) MSTR(text).toCharArray(), MSTR(text).length());
        #endif
    }

/**
*
*/
//% blockId="fs_read_handle" block="read file handle %fd" advanced=true
    int fsRead(int fd) {
    #if MICROBIT_CODAL
        if (fd < 0) return MICROBIT_NOT_SUPPORTED;
        initFileSystem();

        char c[1];
        int ret = uBit.filesystem.read(fd, (uint8_t*)&c, 1);
        if (ret != 1) return ret;
        else return c[0];
        #endif
    }

    /**
    *
    */
    //% blockId="fs_write_buffer" block="write %fd %buffer" advanced=true
    int fsWriteBuffer(int fd, Buffer buffer) {
    #if MICROBIT_CODAL
        if (fd < 0) return MICROBIT_NOT_SUPPORTED;

        initFileSystem();
        return uBit.filesystem.write(fd, buffer->data, buffer->length);
        #endif
    }

    /**
    */
    //% blockId="fs_read_buffer" block="read %fd length %length" advanced=true
    Buffer fsReadBuffer(int fd, int length) {
    #if MICROBIT_CODAL
        if (fd < 0 || length < 0)
            return mkBuffer(NULL, 0);

        initFileSystem();
        Buffer buf = mkBuffer(NULL, length);

        int ret = uBit.filesystem.read(fd, buf->data, buf->length);

        if (ret < 0) return mkBuffer(NULL, 0);
        else if (ret != length) {
            auto sbuf = mkBuffer(buf->data, ret);
            decrRC(buf);
            return sbuf;
        }
        else return buf;
        #endif
    }

    /**
    *
    */
    //% blockId="fs_test_dir" block="create test directories" advanced=true
        void test() {
        #if MICROBIT_CODAL

            initFileSystem();

            uBit.filesystem.createDirectory("makecode-test");
            uBit.filesystem.createDirectory("makecode-test/nested-test");
            uBit.filesystem.createDirectory("makecode-test2");
            uBit.filesystem.createDirectory("makecode-test3");
            uBit.filesystem.createDirectory("makecode-test4");
            #endif
        }

        /**
        *
        */
        //% blockId="fs_format" block="format the filesystem"
            void format() {
            #if MICROBIT_CODAL

                initFileSystem();

                uBit.filesystem.format();
                #endif
            }

}
